CDF documentation update (#273)
IMAP-Science-Operations-Center/imap_processing: docs/source/development/CDFs/cdf_requirements.rst

#####################

> CDF File Requirements

Reviewer: IMAP-Specific CDF Requirements ?

Do we want to be explicit that these are IMAP specific requirements since you say these are necessary for this mission in the next sentence?

#####################

The following page describes the necessary contents inside of a CDF file generated for the IMAP mission.

For a more general introduction to CDF files, see :ref:`CDF Introduction`.

For sample code that will generate a file that matches the below guidelines, see :ref:`IMAP xarray_to_cdf Example`

> ******************************
> Global Attributes
> ******************************

Reviewer 2: ```suggestion
*****************
Global Attributes
*****************
```

Global attributes are used to provide information about the data set as an entity. Together with variables and variable attributes, the global attributes make the data correctly and independently usable by someone not connected with the instrument team, and hence, a good archive product.

Global attributes that have been identified for use with IMAP data products are listed below. Additional Global attributes can be defined but they must start with a letter and can otherwise contain letters, numbers and the underscore character (no other special characters allowed). Note that CDF attributes are case-sensitive and must exactly follow what is shown here.

ISTP Compliant Global Attributes are listed here: `https://spdf.gsfc.nasa.gov/istp_guide/gattributes.html <https://spdf.gsfc.nasa.gov/istp_guide/gattributes.html>`_, and notes about how they are used on IMAP are below -

> Data_type

Reviewer: Would this be easier to follow as a table?
Something like three rows: Attribute Name, Valid Options, Description

It is always a pain to make rst tables though, so this doesn't have to happen now and can be a follow-up if someone does want to do that later.

==========
This attribute is used by CDF file writing software to create a filename. It is a combination of the following filename components: mode, data level, and optional data product descriptor.

> Data_version
> ================

Reviewer 2: ```suggestion
Data_version
============
```

This attribute identifies the version of a particular CDF data file.

> Descriptor
> ================

Reviewer 2: ```suggestion
Descriptor
==========
```

This attribute identifies the name of the instrument or sensor that collected the data.  Both a long name and a short name are given.  For any data file, only a single value is allowed.

For IMAP, one of the following must be used -

* IDEX>Interstellar Dust Experiment
* SWE>Solar Wind Electrons
* SWAPI>Solar wind and Pickup Ions

> * CoDICE>Compact Dual Ion Compoition Experiment

Reviewer 2: ```suggestion
* CoDICE>Compact Dual Ion Composition Experiment
```

* MAG>Magnetometer

> * HIT>High-energy Ion Teleccope

Reviewer 2: ```suggestion
* HIT>High-energy Ion Telescope
```

* GLOWS>GLObal Solar Wind Structure
* IMAP-Hi>Interstellar Mapping and Acceleration Probe High
* IMAP-Lo>Interstellar Mapping and Acceleration Probe Low
* IMAP-Ultra>Interstellar Mapping and Acceleration Probe Ultra

> Discipline
> ================

Reviewer 2: ```suggestion
Discipline
==========
```

For IMAP, this value should always be “Space Physics>Heliospheric Physics.”. This attribute describes both the science discipline and sub discipline.

> Generation_date
> ================

Reviewer 2: ```suggestion
Generation_date
===============
```

Date stamps the creation of the file using the syntax yyyymmdd, e.g. 20150923.

> Instrument_type
> ================

Reviewer 2: ```suggestion
Instrument_type
===============
```

Reviewer 2: Looks like there are several more of these below, so I will stop commenting on them individually 😅

* UNH - University of New Hampshire

PI_name
================
This attribute value should include first initial and last name of the IMAP mission PI followed by a comma-separated list of any Co-Is that are responsible for this particular dataset. For example, a single PI entry in this attribute would be: "Dr. David J. McComas".

Project
================
This attribute identifies the name of the project and indicates ownership. For IMAP, this value should be “STP>Solar-Terrestrial Physics”.

Source_name
================
This attribute identifies the observatory where the data originated. For IMAP, this should simply be "IMAP"

TEXT
================
This attribute is an SPDF standard global attribute, which is a text description of the experiment whose data is included in the CDF. A reference to a journal article(s) or to a World Wide Web page describing the experiment is essential, and constitutes the minimum requirement. A written description of the data set is also desirable. This attribute can have as many entries as necessary to contain the desired information. Typically, this attribute is about a paragraph in length and is not shown on CDAWeb.

MODS
================

> This attribute is an SPDF standard global attribute, which is used to denote the history of modifications made to the CDF data set. The MODS attribute should contain a description of all significant changes to the data set, essentially capturing a log of highlevel release notes. This attribute can have as many entries as necessary and should be updated if the "X" value of the version number changes.

Reviewer 2: ```suggestion
This attribute is an SPDF standard global attribute, which is used to denote the history of modifications made to the CDF data set. The MODS attribute should contain a description of all significant changes to the data set, essentially capturing a log of high-level release notes. This attribute can have as many entries as necessary and should be updated if the "X" value of the version number changes.
```

Parents
================
This attribute lists the parent data files for files of derived and merged data sets. The syntax for a CDF parent is: "CDF>logical_file_id". Multiple entry values are used for multiple parents. This attribute is required for any MMS data products that are derived from 2 or more data sources and the file names of parent data should be clearly identified. CDF parents may include source files with non-cdf extensions.

> ******************************
> IMAP Variables
> ******************************

Reviewer 2: ```suggestion
***************
IMAP Variables
***************
```

There are three types of variables that should be included in CDF files: data, support data, and metadata. Additionally, required attributes are listed with each variable type listed below.

To facilitate data exchange and software development, variable names should be consistent across the IMAP instruments. Additionally, it is preferable that data types are consistent throughout all IMAP data products (e.g. all real variables are CDF_REAL4, all integer variables are CDF_INT4, and flag/status variables are UINT4).
This is not to imply that only these data types are allowable within IMAP CDF files. All CDF supported data types are available for use by IMAP. For detailed information and examples, please see the following ISTP/IACG webpage:
`http://spdf.gsfc.nasa.gov/istp_guide/variables.html <http://spdf.gsfc.nasa.gov/istp_guide/variables.html>`_

Data
==============
These are variables of primary importance (e.g., density, magnetic field, particle flux). Data is always time (record) varying, but can be of any dimensionality or CDF supported data type. Real or Integer data are always defined as having one element.

Required Epoch Variable
------------------------

> All IMAP CDF Data files must contain at least one cariable of data type CDF_TIME_TT2000 named "Epoch".  All time varying variables in the CDF data set will depend on either this "Epoch" or another variable of type CDF_TIME_TT2000.  More than one CDF_TIME_TT2000 variable is allowed in a data set to allow for more than one time resolution.  It is recommended that all such time variable use "Epoch" within their variable name.

Reviewer 2: ```suggestion
All IMAP CDF Data files must contain at least one variable of data type CDF_TIME_TT2000 named "Epoch".  All time varying variables in the CDF data set will depend on either this "Epoch" or another variable of type CDF_TIME_TT2000.  More than one CDF_TIME_TT2000 variable is allowed in a data set to allow for more than one time resolution.  It is recommended that all such time variable use "Epoch" within their variable name.
```

These variable names may only include lower-case letters, numbers, underscores, and hyphens.  No upper-case letter or other special characters are allowed.

Required
---------
* instrument - the instrument acronym
* parameter - a short representation of the physical parameter held in the variable

Optional
---------
* coordinateSystem - an acronym for the coordinate system in which the parameter is set
* parameterQualifier - parameter descriptor, which may include multiple components separated by a "_" as needed (e.g. "pa_0" indicates a pitch angle of 0)
* subModeLevel - Qualifiers to include mode and data level information supplementary to the next to fields
* mode - The mode of the instrument
* dataLevel - The data level in the "Data_type" attribute field

Support Data Variables
=======================

Support data variable names must begin with a letter and can contain numbers and underscores, but no other special characters. Support data variable names need not follow the same naming convention as Data Variables (5.1.1) but may be shortened for convenience.

******************************

> File Naming Convention

Reviewer: I'm wondering if some of this is getting into the `style-guide` territory? I'm not sure where I would expect things to live though.

Reviewer 2: We could add a link to this section the `Naming Conventions` doc in the style guide.
